pkg/scheduler/framework/framework.go

```diff
@@ -143,8 +170,161 @@ func (f *framework) EventRecorder() record.EventRecorder {
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Scheduling cycle ends", "clusterSchedulingPolicySnapshot", policyRef, "latency", latency)
+	}()
+
+	// TO-DO (chenyu1): add metrics.
+
+	// Collect all clusters.
+	//
+	// Note that clusters here are listed from the cached client for improved performance. This is
+	// safe in consistency as it is guaranteed that the scheduler will receive all events for cluster
+	// changes eventually.
+	clusters, err := f.collectClusters(ctx)
+	if err != nil {
+		klog.ErrorS(err, "failed to collect clusters", "clusterSchedulingPolicySnapshot", policyRef)
+		return ctrl.Result{}, err
+	}
+
+	// Collect all bindings.
+	//
+	// Note that for consistency reasons, bindings are listed directly from the API server; this helps
+	// avoid a classic read-after-write consistency issue, which, though should only happen when there
+	// are connectivity issues and/or API server is overloaded, can lead to over-scheduling in adverse
+	// scenarios. It is true that even when bindings are over-scheduled, the scheduler can still correct
+	// the situation in the next cycle; however, considering that placing resources to clusters, unlike
+	// pods to nodes, is more expensive, it is better to avoid over-scheduling in the first place.
+	//
+	// This, of course, has additional performance overhead (and may further exacerbate API server
+	// overloading). In the long run we might still want to resort to a cached situtation.
+	//
+	// TO-DO (chenyu1): explore the possbilities of using a mutation cache for better performance.
+	bindings, err := f.collectBindings(ctx, crpName)
+	if err != nil {
+		klog.ErrorS(err, "failed to collect bindings", "clusterSchedulingPolicySnapshot", policyRef)
+		return ctrl.Result{}, err
+	}
+
+	// Parse the bindings, find out
+	//
+	// * bound bindings, i.e., bindings that are associated with a normally operating cluster and
+	//   have been cleared for processing by the dispatcher; and
+	// * scheduled bindings, i.e., bindings that have been associated with a normally operating cluster,
+	//   but have not yet been cleared for processing by the dispatcher; and
+	// * obsolete bindings, i.e., bindings that are scheduled in accordance with an out-of-date
+	//   (i.e., no longer active) scheduling policy snapshot; it may or may have been cleared for
+	//   processing by the dispatcher; and
+	// * dangling bindings, i.e., bindings that are associated with a cluster that is no longer
+	//   in a normally operating state (the cluster has left the fleet, or is in the state of leaving),
+	//   yet has not been marked as unscheduled by the scheduler; and
+	//
+	// Note that bindings marked as unscheduled are ignored by the scheduler, as they
+	// are irrelevant to the scheduling cycle. Any deleted binding is also ignored.
+	bound, scheduled, obsolete, dangling := classifyBindings(policy, bindings, clusters)
```

Comment:
how about creating a separate struct to hold all the bindings and right now they're used in multiple place,

something like,

type ClassifiedBindings struct {
       Bound []*fleetv1beta1.ClusterResourceBinding
       Scheduled []*fleetv1beta1.ClusterResourceBinding
       Obsolete []*fleetv1beta1.ClusterResourceBinding
       Dangling []*fleetv1beta1.ClusterResourceBinding
}

Rollout controller could also reuse this struct.

Reply:
I don't think that's really making the code more readable.

`markAsDeletingFor(..., classifiedBindings.Dangling)` is no better than `markAsDeletingFor(..., dangling)`; and passing the whole struct in is even more confusing: `markAsDeletingFor(..., classifiedBindings)`

Reply:
oh, i was thinking using this structure in the runSchedulingCycleForPickNPlacementType & runSchedulingCycleForPickAllPlacementType, which have long input arg list.

Later if we have more binding state, we don't need to update the func interface.